### PR TITLE
feat(hc): Use hybrid cloud services in AuthIdentityHandler

### PR DESCRIPTION
**Pull request by RyanSkonnord**
These changes are targeted toward getting all AuthHelper and AuthIdentityHandler code to run with control silo limits. Most availability errors from AuthIdentityHandler are fixed, but we need more work around the Actor model and ApiInviteHelper to make it stable.

### CODE REVIEW

File: src/sentry/auth/helper.py [hunk 1]
```diff
@@ -98,10 +105,15 @@ class AuthIdentityHandler:
...
     auth_provider: AuthProvider
     provider: Provider
-    organization: Organization
+    organization: ApiOrganization
     request: HttpRequest
     identity: Mapping[str, Any]
 
+    def __post_init__(self) -> None:
+        # For debugging. TODO: Remove when tests are stable
+        if not isinstance(self.organization, ApiOrganization):
```
Comment (corps) [inline src/sentry/auth/helper.py:114]: 👍 
You could also make this conditional to tests by 

```
import sys
if 'pytest' in sys.modules:
   ...
```

Which means you can safely continue to leave it.

File: src/sentry/auth/helper.py [hunk 2]
```diff
@@ -222,54 +234,25 @@ def _get_login_redirect(self, subdomain: str | None) -> str:
...
-        # Otherwise create a new membership
-        om = OrganizationMember.objects.create(
-            organization=self.organization,
-            role=self.organization.default_role,
-            user=user,
-            flags=flags,
+    def _handle_new_membership(
+        self, auth_identity: ApiAuthIdentity
+    ) -> ApiOrganizationMember | None:
+        user, om = auth_service.handle_new_membership(
```
Comment (corps) [inline src/sentry/auth/helper.py:240]: 👍 Nice.

File: src/sentry/auth/helper.py [hunk 3]
```diff
@@ -337,16 +320,19 @@ def handle_attach_identity(self, member: OrganizationMember | None = None) -> Au
 
         if member is None:
             member = self._get_organization_member(auth_identity)
-        self._set_linked_flag(member)
+        if member is not None:
```
Comment (corps) [inline src/sentry/auth/helper.py:323]: Hmm, I'm curious -- why the new conditional? Was member previous always not None but now is, or did the set_linked_flag just gracefully handle None?

File: src/sentry/auth/helper.py [hunk 4]
```diff
@@ -750,8 +738,14 @@ def finish_pipeline(self) -> HttpResponseBase:
         return response
 
     def auth_handler(self, identity: Mapping[str, Any]) -> AuthIdentityHandler:
+        # This is a temporary step to keep test_helper integrated
+        # TODO: Move this conversion further upstream
+        api_organization = DatabaseBackedOrganizationService.serialize_organization(
```
Comment (corps) [inline src/sentry/auth/helper.py:743]: 👍

File: src/sentry/services/hybrid_cloud/audit/impl.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        if user is None:
+            return None
+        if isinstance(user, int):
+            user_id = user
+        else:
+            user_id = user.id
+        return User.objects.get(id=user_id)
+
+    def write_audit_log(self, *, metadata: AuditLogMetadata, data: Mapping[str, Any]) -> None:
+        organization = Organization.objects.get(id=metadata.organization.id)
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/audit/impl.py:24]: How is this working?  I'm pretty sure as written, this database backed service must run on `control`, which means this shouldn't be able to fetch an organization object.

Would it also not suffice to simple capture the organization id and pass it to the `create`?

Basically, we cannot realistically have an Organization ORM lookup in the same logic that writes an AuditLogEntry.

File: src/sentry/services/hybrid_cloud/audit/impl.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        AuditLogEntry.objects.create(
+            organization=organization,
+            event=metadata.event,
+            actor_label=metadata.actor_label,
+            actor=self._resolve_user(metadata.actor),
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/audit/impl.py:29]: Does this need the full actor object?  Would it be sufficient to pass an id alone?  Also @vbro for 👀

File: src/sentry/services/hybrid_cloud/audit/impl.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            target_object=metadata.target_object,
+            target_user=self._resolve_user(metadata.target_user),
+            data=data,
+        )
+
+    def log_organization_membership(
+        self, *, metadata: AuditLogMetadata, organization_member: ApiOrganizationMember
+    ) -> None:
+        user_id = organization_member.user_id
+        assert user_id is not None
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/audit/impl.py:40]: Hmmm... I think this is wrong.  Just because a user does not exist, does not mean we cannot log something about their id.  Many AuditLogEntry invocations come from async processes, where technically they can have a user_id in hand of a user that has already been deleted.  At the moment, we use cascade logic to `set_null` those user_ids, but in the future an async process will provide that. 

I think the correct thing is to take the user_id provided "as is" without validating its deletion and allow eventually consistency of the tombstone / outbox pattern I'm working on to handle the rest.

So basically, *we still want audit log entries even when a user no longer exists*, and *we'll handle the set null at a different level*.

File: src/sentry/services/hybrid_cloud/audit/impl.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        user = user_service.get_user(user_id=user_id)
+        assert user is not None
+
+        team_ids = [mt.team_id for mt in organization_member.member_teams]
+        team_slugs = list(Team.objects.filter(id__in=team_ids).values_list("slug", flat=True))
+
+        data = {
+            "email": user.email,
+            "user": None,  # TODO,
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/audit/impl.py:49]: I think it's sufficient to swap this for `"user_id": metadata.user.id`

File: src/sentry/services/hybrid_cloud/audit/impl.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+from typing import Any, Mapping
+
+from sentry.models import AuditLogEntry, AuthIdentity, AuthProvider, Organization, Team, User
+from sentry.services.hybrid_cloud.audit import AuditLogMetadata, AuditLogService
+from sentry.services.hybrid_cloud.auth import ApiAuthIdentity, ApiAuthProvider
+from sentry.services.hybrid_cloud.organization import ApiOrganizationMember
+from sentry.services.hybrid_cloud.user import APIUser, user_service
+
+
+class DatabaseBackedAuditLogService(AuditLogService):
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/audit/impl.py:12]: A high level concern: right now, I actually already did implement an AuditLogEntry service:

```
    def save_or_write_to_kafka(self):
        """
        Region Silos do not have access to the AuditLogEntry table which is specific to the control silo.
        For those silos, this method publishes the attempted audit log write to a durable kafka queue synchronously
        that will eventually be consumed by the control silo.  For the control silo, this method ultimately results
        in a save() call.
        """
        from sentry.region_to_control.producer import audit_log_entry_service

        audit_log_entry_service.produce_audit_log_entry(self)
```

Which is hooked up to some existing call sites, but isn't running in production.  However, it is currently used in a handful of tests.

One major difference is that in my original design, AuditLogs were written to kafka and asynchronously written to the database.  I do think we want to move *away from kafka*, but I wonder if using Outbox to synchronize these might also be a strong option.  It would improve latency and availability of these code paths since writing an AuditLog is never a synchronously required action.

Strictly speaking, such work is not necessary for this PR for now (let's get the auth login work across the line), but when the login work is wrapping up, let's think about how we can converge on a single audit log service, and wether we think it should remain synchronous or not.

File: src/sentry/services/hybrid_cloud/auth/__init__.py [hunk 10]
```diff
@@ -134,6 +134,16 @@ def get_auth_providers(self, organization_id: int) -> List[ApiAuthProvider]:
         """
         pass
 
+    @abc.abstractmethod
+    def handle_new_membership(
+        self,
+        request: Request,
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/auth/__init__.py:140]: Hmm, my worry here is the requirement of the request object.  Is there, any way we can further unpack this?  Because when this becomes an RPC invocation, we don't want to be serializing an entire request object.

File: src/sentry/services/hybrid_cloud/auth/impl.py [hunk 11]
```diff
@@ -240,6 +257,64 @@ def get_org_ids_with_scim(
...
+            invite_helper.handle_invite_not_approved()
+
+        flags = ApiOrganizationMemberFlags(sso__linked=True)
+        # if the org doesn't have the ability to add members then anyone who got added
+        # this way should be disabled until the org upgrades
+        if not features.has("organizations:invite-members", orm_org):
+            flags.member_limit__restricted = True
+
+        # Otherwise create a new membership
+        om = organization_service.add_organization_member(
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/auth/impl.py:302]: Cool.  In the future, this will become simpler: we'll create an `OrganizationMemberMapping` orm locally on the control, and that will synchronize back.  But this is sufficient for current world compatibility.

**Review by corps — COMMENTED** [review]
Cool!  I think, as you said at standup, taking the approach by *starting with setting test to stable=True* and following the implications that surfaces is a great way to make progress.  I think this is a very smart way to break up the complex work and I can see where you're headed.  I added a lot of feedback but I want to emphasize that I agree with your approach and am grateful for breaking this down.

File: src/sentry/auth/helper.py (hunk 3 above)
Comment (RyanSkonnord) [reply src/sentry/auth/helper.py]: I had declared `AuthService.handle_new_membership` to return an optional membership at some point but no longer needs to. That trickled down to here and I added the null check assuming the contract had changed. Fixed in 09b79a9a645b716b6d41a1b06b43b8225b80e777

File: src/sentry/services/hybrid_cloud/audit/impl.py (hunk 5 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/audit/impl.py:24]: That's a mistake on my part. I believe it would have shown up running `AuthHelperTest` with `stable=True`, if it weren't being masked by a different availability error that's already coming from `ApiInviteHelper`. I'll be fixing it as part of getting those tests to pass. In the meantime, we can treat it like https://github.com/getsentry/sentry/blob/09b79a9a645b716b6d41a1b06b43b8225b80e777/src/sentry/services/hybrid_cloud/auth/impl.py#L272-L274

File: src/sentry/services/hybrid_cloud/audit/impl.py (hunk 6 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/audit/impl.py:29]: That sounds right. I'll experiment.

File: src/sentry/services/hybrid_cloud/audit/impl.py (hunk 7 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/audit/impl.py:40]: Sounds good. I would guess that, when I dropped in that assertion, I was thinking more about getting Mypy to pass than about the correct behavior for audit logs.

It also wouldn't surprise me if, before my changes, the rare case of a null user would have simply caused an error, and it's just more obvious now.

File: src/sentry/services/hybrid_cloud/audit/__init__.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def log_organization_membership(
+        self,
+        *,
+        metadata: AuditLogMetadata,
+        organization_member: ApiOrganizationMember,
+    ) -> None:
+        raise NotImplementedError
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/audit/__init__.py:42]: Are you thinking that there will be a method for each audit log that is recorded? We have quite a few audit log event types which will make this interface wide.

We should be able to extend the existing AuditLogEvent plumbing code to be aware of required data attributes which would let us have an interface like

```python
audit_log_service.log(AuditLogMetaData( 
    organization=self.organization,
    actor=self.user,
    ip_address=self.request.META["REMOTE_ADDR"],
    target_object=auth_identity.id,
    event=audit_log.get_event_id("SSO_IDENTITY_LINK"),
    auth_identity=auth_identity
)
```
With required parameters embedded in to the `event` definition `AuditLogMetaData` could ensure all the required attributes are present. This approach would have a narrow interface, but wouldn't be as typesound as what you have presently.

File: src/sentry/services/hybrid_cloud/auth/__init__.py (hunk 10 above)
Comment (markstory) [reply src/sentry/services/hybrid_cloud/auth/__init__.py:140]: It looks like we rely on two session values, `invite_token` and `invite_member_id` and the request user.

File: src/sentry/services/hybrid_cloud/organization/impl.py [hunk 13]
```diff
@@ -235,3 +251,41 @@ def _query_organizations(
...
+        )
+        return self.serialize_member(member)
+
+    def add_team_member(
+        self, *, team_id: int, organization_member: ApiOrganizationMember
+    ) -> ApiTeamMember:
+        omt = OrganizationMemberTeam.objects.create(
+            team_id=team_id, organizationmember_id=organization_member.id
+        )
+        project_ids = ()  # TODO?
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/organization/impl.py:281]: Should project_ids be a parameter?

File: src/sentry/services/hybrid_cloud/auth/__init__.py (hunk 10 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/auth/__init__.py:140]: Sounds good, but it would cascade some sizeable changes through `ApiInviteHelper`. How about as a separate PR?

Comment (markstory) [reply src/sentry/services/hybrid_cloud/auth/__init__.py:140]: Totally makes sense to do separately :+1:

Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/auth/__init__.py:140]: https://github.com/getsentry/sentry/pull/43211

File: src/sentry/services/hybrid_cloud/organization/impl.py (hunk 13 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/organization/impl.py]: Per discussion with @corps, we'll just sidestep this by not having a return value: 3c3e5538d8761cd6ab559e74ba0a9eb4dac9cd13

**Comment by RyanSkonnord** [comment]
Regarding several comments on and near the writing of audit logs: I agree with your feedback, and everything around the audit logs was pretty much a hack job trying to get this integrated with minimal code movement outside of `AuditLogService`. Per discussion with @corps, I think the most realistic way to ship these auth changes would be to punt the audit log stuff until we can cover it with a good general mechanism.

**Review by corps — APPROVED** [review]
